cas-server-support-generic/src/main/java/org/jasig/cas/adaptors/generic/ShiroAuthenticationHandler.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed to Apereo under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work
+ * for additional information regarding copyright ownership.
+ * Apereo licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License.  You may obtain a
+ * copy of the License at the following location:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.jasig.cas.adaptors.generic;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.shiro.SecurityUtils;
+import org.apache.shiro.authc.AuthenticationException;
+import org.apache.shiro.authc.DisabledAccountException;
+import org.apache.shiro.authc.ExcessiveAttemptsException;
+import org.apache.shiro.authc.ExpiredCredentialsException;
+import org.apache.shiro.authc.IncorrectCredentialsException;
+import org.apache.shiro.authc.LockedAccountException;
+import org.apache.shiro.authc.UnknownAccountException;
+import org.apache.shiro.authc.UsernamePasswordToken;
+import org.apache.shiro.config.IniSecurityManagerFactory;
+import org.apache.shiro.mgt.SecurityManager;
+import org.apache.shiro.subject.Subject;
+import org.apache.shiro.util.Factory;
+import org.jasig.cas.authentication.AccountDisabledException;
+import org.jasig.cas.authentication.Credential;
+import org.jasig.cas.authentication.HandlerResult;
+import org.jasig.cas.authentication.PreventedException;
+import org.jasig.cas.authentication.RememberMeUsernamePasswordCredential;
+import org.jasig.cas.authentication.UsernamePasswordCredential;
+import org.jasig.cas.authentication.handler.support.AbstractUsernamePasswordAuthenticationHandler;
+import org.springframework.core.io.Resource;
+
+import javax.security.auth.login.AccountLockedException;
+import javax.security.auth.login.AccountNotFoundException;
+import javax.security.auth.login.CredentialExpiredException;
+import javax.security.auth.login.FailedLoginException;
+import java.security.GeneralSecurityException;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * An authentication handler that routes requests to Apache Shiro.
+ * Credentials are assumed to be username and password.
+ * @author Misagh Moayyed
+ * @since 4.2
+ */
+public class ShiroAuthenticationHandler extends AbstractUsernamePasswordAuthenticationHandler {
```

Comment:
One possible advantage might be that CAS would become a comprehensive SSO platform to not only work with a variety of apps, but also provide support for a variety of well adopted frameworks, so that we could cross this off our list. Similar to JWT, Mongo, and Stormpath, "if you build it, they will come". Specially for Shiro, JWT and Mongo, I have had several questions on whether CAS does this sort of thing. So we should keep moving fwd.